### PR TITLE
Immediately call `onCancel` handlers when a promise is canceled

### PR DESCRIPTION
**Pull request by jonschlinkert**
Fixes https://github.com/sindresorhus/p-cancelable/issues/15

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -46,7 +48,12 @@ class PCancelable {
 					throw new Error('The `onCancel` handler was attached after the promise settled.');
 				}
 
-				this._cancelHandlers.push(handler);
+				const fn = (...args) => {
```
Comment (szmarczak) [inline index.js:51]: Can you undo this change? A much better solution is to move the `this._isCanceled = true;` before the if here:

https://github.com/jonschlinkert/p-cancelable/blob/deaeee112adc3d105aaec2074de971af5dd48c3a/index.js#L90-L100

File: index.js [hunk 2]
```diff
@@ -32,8 +32,10 @@ class PCancelable {
 			this._reject = reject;
 
 			const onResolve = value => {
-				this._isPending = false;
-				resolve(value);
+				if (this._isCanceled !== true || onCancel.shouldReject === false) {
```
Comment (szmarczak) [inline index.js:35]:
```suggestion
				if (!this._isCanceled || !onCancel.shouldReject) {
```

**Comment by szmarczak** [comment]
Add a test:

```js
test('throws immediately as soon as .cancel() is called', async t => {
	const cancelablePromise = new PCancelable((resolve, reject, onCancel) => {
	const timeout = setTimeout(() => {
		resolve(true);
	}, 10);
	
	onCancel.shouldReject = true;
	onCancel(() => {
		clearTimeout(timeout);
		resolve(false);
	  });
	});

	cancelablePromise.cancel();

	await t.throwsAsync(cancelablePromise, {
		message: 'Promise was canceled'
	});
});
```

File: index.js (hunk 1 above)
Comment (sindresorhus) [reply index.js:51]: https://github.com/sindresorhus/p-cancelable/commit/4b7d014c5be44889e40014dea098ae44819b95d6

**Comment by ifiokjr** [comment]
Is this still being worked on?

**Comment by szmarczak** [comment]
@ifiokjr That would be great! I don't think the author is active on GitHub anymore (I just guess by looking at his contributions chart).

**Comment by sindresorhus** [comment]
Friendly bump :)

**Comment by jonschlinkert** [comment]
Oh wow, I totally forgot. Thanks for the ping.

**Comment by sindresorhus** [comment]
Thanks :)
